backend/msgpack/placement.go

```diff
@@ -0,0 +1,269 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	schema "github.com/m3db/m3cluster/generated/proto/placement"
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/clock"
+)
+
+var (
+	errNilPlacementSchema            = errors.New("nil placement schema")
+	errNilPlacementSnapshotsSchema   = errors.New("nil placement snapshots schema")
+	errNoApplicablePlacementSnapshot = errors.New("no applicable placement snapshot found")
+	errPlacementClosed               = errors.New("placement is closed")
+)
+
+// placement stores placement information.
+type placement interface {
+	// NumShards returns the number of shards.
+	NumShards() (int, error)
+
+	// Shard returns the shard associated with a given metric id.
+	Shard(id []byte) (uint32, error)
+
+	// RouteShard routes msgpack encoded buffer for a given shard.
+	RouteShard(shard uint32, buf msgpack.Buffer) error
+
+	// Close closes the topology.
+	Close() error
+}
+
+type activePlacement struct {
+	sync.RWMutex
+
+	snapshots []*placementSnapshot
+	nowFn     clock.NowFn
+	queueMgr  instanceQueueManger
+	expiring  int32
+	closed    bool
+}
+
+func newActivePlacement(snapshots []*placementSnapshot, opts placementOptions) placement {
+	queueMgr := opts.InstanceQueueManger()
+	for _, snapshot := range snapshots {
+		queueMgr.AddInstances(snapshot.Instances())
+	}
+	return &activePlacement{
+		snapshots: snapshots,
+		nowFn:     opts.ClockOptions().NowFn(),
+		queueMgr:  queueMgr,
+	}
+}
+
+func (p *activePlacement) NumShards() (int, error) {
+	p.RLock()
+	snapshot, err := p.activeSnapshotWithLock(p.nowFn())
+	if err != nil {
+		p.RUnlock()
+		return 0, err
+	}
+	numShards := snapshot.NumShards()
+	p.RUnlock()
+	return numShards, nil
+}
+
+func (p *activePlacement) Shard(id []byte) (uint32, error) {
+	p.RLock()
+	snapshot, err := p.activeSnapshotWithLock(p.nowFn())
+	if err != nil {
+		p.RUnlock()
+		return 0, err
+	}
+	shard := snapshot.Shard(id)
+	p.RUnlock()
+	return shard, nil
+}
+
+func (p *activePlacement) RouteShard(shard uint32, buf msgpack.Buffer) error {
+	p.RLock()
```

Comment:
While it would be a rare race, we have encountered these sorts of race conditions in other systems when changing topology in the past.

I'm not too sure whether this belongs more with topology or placement, but I imagine an interface with route(buf msgpack.Buffer) that is not racey would be ideal for users of this. They shouldn't even care that the message is routed via uint32 shards. I can imagine in the future, we want to change the sharding strategy to something that's not just a uint32 (combination of cluster for various types/groups of metrics for example)